Feat: Implement proxy (#1)
open-runtimes/proxy: README.md

# Open Runtimes Proxy 🚦

![open-runtimes-box-bg-cover](https://user-images.githubusercontent.com/1297371/151676246-0e18f694-dfd7-4bab-b64b-f590fec76ef1.png)

---

[![Discord](https://img.shields.io/discord/937092945713172480?label=discord&style=flat-square)](https://discord.gg/mkZcevnxuf)

> [![Build Status](https://github.com/utopia-php/balancing/actions/workflows/tester.yml/badge.svg)](https://github.com/utopia-php/balancing/actions/workflows/tester.yml)

Reviewer: Replace this link with that for the proxy

[![Twitter Account](https://img.shields.io/twitter/follow/appwrite?color=00acee&label=twitter&style=flat-square)](https://twitter.com/appwrite)

> <!-- [![Docker Pulls](https://img.shields.io/docker/pulls/appwrite/appwrite?color=f02e65&style=flat-square)](https://hub.docker.com/r/appwrite/appwrite) -->

Reviewer: You can use the link to the proxy docker hub image

Author: It will only work after we release first version. Added it.

Proxy server for [Open Runtimes](https://github.com/open-runtimes/open-runtimes), a runtime environments for serverless cloud computing for multiple coding languages.

The proxy is responsible for checking health of executors, and proxying requests between them based on selected strategy, such as round-robin. Proxy is stateless and can be scaled horizontally when a load balanced is introduced in front of it.

## Features

* **Flexibility** - Adapter-driven approach allows switching between proxying strategies. You can even implement your own.
* **Performance** - Coroutine-style HTTP servers allows asynchronous operations without blocking. We. Run. Fast! ⚡
* **Open Source** - Released under the MIT license, free to use and extend.

## Getting Started - Open Runtimes Executor

> > 🚧 Getting started with Open Runtimes Executor coming soon.

Reviewer: Why do we need this section at all ? Shouldn't this be in the readme for the executor ?

Author: Good point, you would never know that proxy exists this way. We should have this in the executor readme.

Author: removing it from here.

## Getting Started - Generic Proxy

1. Pull Open Runtimes Proxy image:

```bash
docker pull openruntimes/proxy
```

2. Pull whoami image:

```bash
docker pull containous/whoami
```

3. Create `docker-compose.yml` file:

```yml
version: '3'
services:
  openruntimes-proxy:
    image: openruntimes/proxy
    ports:
      - 9800:80
    environment:
      - OPEN_RUNTIMES_PROXY_ALGORITHM
      - OPEN_RUNTIMES_PROXY_EXECUTORS

>       - OPEN_RUNTIMES_PROXY_PING_INTERVAL

Reviewer 2: ```suggestion
      - OPEN_RUNTIMES_PROXY_HEALTHCHECK_INTERVAL
```

Could this be a better name?

|-----------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|
| OPEN_RUNTIMES_PROXY_ALGORITHM           | Proxying algorithm. Supports `round-robin`, `random`.                                                                                     |
| OPEN_RUNTIMES_PROXY_EXECUTORS           | Comma-separated hostnames of servers under the proxy.                                                                                     |
| OPEN_RUNTIMES_PROXY_HEALTHCHECK | Health check by HTTP request to /v1/health. 'enabled' by default. To disable, set to 'disabled'.                                          |
| OPEN_RUNTIMES_PROXY_HEALTHCHECK_INTERVAL       | Delay in milliseconds between health checks. 10000 by default. Only relevant if OPEN_RUNTIMES_PROXY_HEALTHCHECK is 'enabled'.     |
| OPEN_RUNTIMES_PROXY_ENV                 | Runtime environment. 'production' or 'development'. Development may expose debug information and is not recommended on production server. |

> | OPEN_RUNTIMES_PROXY_SECRET              | Secret that needs to be provided in `Authroization` header when talking to proxy.                                                         |

Reviewer: ```suggestion
| OPEN_RUNTIMES_PROXY_SECRET              | Secret that needs to be provided in `Authroization` header when communicating with the proxy.                                                         |
```

| OPEN_RUNTIMES_PROXY_EXECUTOR_SECRET     | String provided as `Authorization` header by proxy when sending request to executor.                                                      |

> | OPEN_RUNTIMES_PROXY_LOGGING_PROVIDER    | Logging provider. Supports `sentry`, `appsignal`, `raygun`, `logown`. Leave empty for no cloud logging.                                   |

Reviewer: ```suggestion
| OPEN_RUNTIMES_PROXY_LOGGING_PROVIDER    | Logging provider. Supports `sentry`, `appsignal`, `raygun`, `logowl`. Leave empty to disable errro logging.                                   |
```

Author: It does not fully disable logging. You still get logging into console. Logs are just not reported to cloud providers.

[![Build Status](https://github.com/utopia-php/balancing/actions/workflows/tester.yml/badge.svg)](https://github.com/utopia-php/balancing/actions/workflows/tester.yml)
[![Twitter Account](https://img.shields.io/twitter/follow/appwrite?color=00acee&label=twitter&style=flat-square)](https://twitter.com/appwrite)

<!-- [![Docker Pulls](https://img.shields.io/docker/pulls/appwrite/appwrite?color=f02e65&style=flat-square)](https://hub.docker.com/r/appwrite/appwrite) -->

Proxy server for [Open Runtimes](https://github.com/open-runtimes/open-runtimes), a runtime environments for serverless cloud computing for multiple coding languages.

The proxy is responsible for checking health of executors, and proxying requests between them based on selected strategy, such as round-robin. Proxy is stateless and can be scaled horizontally when a load balanced is introduced in front of it.

## Features

* **Flexibility** - Adapter-driven approach allows switching between proxying strategies. You can even implement your own.
* **Performance** - Coroutine-style HTTP servers allows asynchronous operations without blocking. We. Run. Fast! ⚡

> * **Open Source** - Released under the MIT license, free to use and extend.

Reviewer 2: We should add documentation to env vars and other relevant stuff that would usually be covered as Appwrite docs.
